WIP: SOSS with XTypes

lemunozm opened the pull request:
In this PR we added the *xtypes* functionality to *soss*.

Things to take into account:
* We added another package as an example of how to build a system handle (especially focused on the use of *soss* with *xtypes*). The idea is to let this package grow with ever new features coming from *xtypes* module, and if wanted, keep it clean allowing the user to take it as a template for other system handles.

Currently the following packages have been updated:
- Internal packages:
  - [x] core
  - [x] mock
  - [x] echo (formerly called local-example)
  - [x] ros2 and ros2-test 
  - [x] json
  - [x] websocket and websocket-test.
- External packages (from osrf):
  - [ ] ros1 and ros1-test
- External packages (from eProsima):
  - [x] fiware and fiware-test (https://github.com/eProsima/SOSS-FIWARE/pull/17)
  - [x] dds and dds-test (https://github.com/eProsima/SOSS-DDS/pull/24)
- Other features
  - [x] YAML configuration file with IDL. 

Please, feel free to give us feedback about the incorporation of *xtypes* to *soss*.
I will update this PR description with links to the documentation and examples:
- [xtypes API documentation](https://github.com/eProsima/xtypes)
- [xtypes API usage example](https://github.com/eProsima/xtypes/blob/master/examples/complex_type.cpp)
- [System Handle conversion code example](https://github.com/eProsima/soss_v2/blob/feature/xtypes-dds/packages/json/src/conversion.cpp) (updated)
- [Functional QoS round trip example: Vec2 -> Vec3 -> Vec3 -> Vec2](https://github.com/eProsima/soss_v2/tree/feature/xtypes-dds/packages/echo/example)

lemunozm committed:
Core compilation with xtypes.

lemunozm committed:
Added idl->xtypes model

lemunozm committed:
xtypes at configure time.

lemunozm committed:
DynamicData passed as a pointer.

lemunozm committed:
Mock compilation.

lemunozm committed:
Minnor fixes.

lemunozm committed:
Added example (first approach)

lemunozm committed:
Working example.

lemunozm committed:
type checks with equality.

lemunozm committed:
global types map to local and checked all compatibility cases at route level.

lemunozm committed:
Addapt mock and xtypes-example to core changes

lemunozm committed:
Fixed checks.

lemunozm committed:
Renamed DynamicType/Data by MessageType/Data

lemunozm committed:
Fixed error. MessageData pointer to references.

lemunozm committed:
Added test structure for xtypes-example. Improved xtypes example. Fixed check compatiblities.

lemunozm committed:
publish tests working.

lemunozm committed:
Fixed subscription test.

lemunozm committed:
Fixed bug at remap function.

lemunozm committed:
Modify guard names.

lemunozm committed:
Added thirdparties

lemunozm committed:
Added fastrtps compilation.

lemunozm committed:
Fixed fastrtps installation.

lemunozm committed:
soss-core compilation with dds-xtypes.

lemunozm committed:
soss-mock compilation with dds-xtypes

lemunozm committed:
xtypes-example adapted to dds-xtypes.

lemunozm committed:
Updated fast. Better configure example.

lemunozm committed:
soss-core compiling with 'xtype parameter'

lemunozm committed:
soss-mock compiling with 'type parameter'

lemunozm committed:
soss-xtypes-example compiling with 'type parameter'

lemunozm committed:
Core and Mock working with xtypes.

lemunozm committed:
xtypes-example module working with basic examples.

lemunozm committed:
soss + xtypes working (examples and test)

lemunozm committed:
Using added. Upgrading document.

lemunozm committed:
Added type registry access from InstanceHandle

lemunozm committed:
Rename to local-example.

lemunozm committed:
Update xtypes and related doc.

lemunozm committed:
Updated fastrpts thirdparty.

lemunozm committed:
Prepared QoS policies. Updated xtypes module.

lemunozm committed:
QoS policies working at example.

lemunozm committed:
Updated xtypes: WString support.

marcoag commented:
It seems that this brings a dependency on dds (`#include <dds/core/xtypes/xtypes.hpp>`). I understand the final objective of this is to create an independent package for xtypes. If that is the case how about we merge this into a branch (i.e. soss_v3) in the meantime until the packaging work is completed?

lemunozm commented:
> It seems that this brings a dependency on dds (`#include <dds/core/xtypes/xtypes.hpp>`). I understand the final objective of this is to create an independent package for xtypes. If that is the case how about we merge this into a branch (i.e. soss_v3) in the meantime until the packaging work is completed?

Hi @marcoag, 

At this moment, the dependency is managed as a thirdparty. When we reach a stable version, this dependency could be managed as an external project (as `yamlcpp` is managed currently in `soss_v2`). 
Anyway, we are open to new ideas.

marcoag commented on .gitmodules:
```diff
@@ -0,0 +1,9 @@
+	url = https://github.com/eProsima/foonathan_memory_vendor.git
+[submodule "thirdparty/fastcdr"]
+	path = thirdparty/fastcdr
+	url = https://github.com/eProsima/Fast-cdr.git
```
afaik git submodules tend to be error prone as they require some good git knowledge and discipline, specially when trying to keep versions between branches synced, any alternative to this we can think of?

lemunozm replied:
That's true, the actual use of *git submodules* is temporal. We are analysing the idea to manage these dependencies with the *superbuild* pattern of `cmake`.

marcoag replied:
I see... is all the code in those three repos needed for the integration of xtypes? would it be an option to somehow extract the relevant parts of some specific stable version and move them into thirdparty?

lemunozm replied:
*xtypes* could be separated in a future. At this moment, we have the library inside because we are developing `fastrtps` over it, and xtypes is an integral part of it. An advantage of it is that as `ros2` use `fastrtps` inside, the `soss-ros2` system handle could avoid the conversion phase because its native message can propagate to soss directly. The same can be applied to `soss-dds` system handle.

Anyway, can be a topic of discussion for the next meeting.

marcoag replied:
I believe if we could have separate code for `xtypes` would be the best as `soss` is not (and shouldn't be) necessarily bounded to`ros2`, so it's dependency on `fastrtps` just for `xtypes` becomes a bit of an overkill solution imho.

lemunozm replied:
*xtypes* is mainly inside of *fastrtps* because from the *fastrtps* view it is a core part and the  *fastrtps* development is easier with this module inside. Anyway, we understand perfectly the *soss* view in this regard. We are currently working in to extract *xtypes* from *fastrtps* to offer this feature without their dependencies. I hope in the next few days to offer a SOSS commit only with *xtypes* dependecy.

marcoag replied:
That sounds great, thanks!

lemunozm replied:
The newly added commit uses *xtypes* as the only third-party. In the future, and in order to avoid git management of submodules, we can add it as a *CMake ExternalProject* and download it automatically.

marcoag replied:
Yup, that sounds like a good idea.

marcoag commented on doc/soss-xtypes-upgrade.md:
````diff
@@ -0,0 +1,238 @@
+
+
+### `configure` function
+Previous
+```c++
+bool configure(
+    const RequiredTypes& types,
+    const YAML::Node& configuration) = 0;
+```
+Now
````
This applies to all the "previous"  on the md file:
 
Do we really need to keep the "previous" way of doing it? is it still supported?

lemunozm replied:
The "previous" way is not supported. This document is only to explain the soss transition to xtypes, or to help some user to adapt its SystemHandle for working with xtypes.

marcoag replied:
Ok, let's keep it for now as a reference then

marcoag commented on packages/core/include/soss/Instance.hpp:
```diff
@@ -74,6 +76,12 @@ class InstanceHandle
+  /// \brief Request the TypeRegitry for a middleware
+  ///
+  /// \returns a pointer to the TypeRegistry or nullptr if the middleware does
+  /// not exits.
```
exit or exist?

lemunozm replied:
You are right, I will modify it.

marcoag commented on packages/core/include/soss/Message.hpp:
```diff
@@ -18,104 +18,8 @@
-#include <map>
-#include <memory>
-#include <string>
+#include <dds/core/xtypes/xtypes.hpp>
```
Is this file still needed or can we move this include to the relevant files?

lemunozm replied:
The purpose of `dds/core/xtypes/xtypes.hpp` is to bring all the necessary of xtypes to the soss ecosystem.

I do not know if you are referring to make separated *includes*, (one for each type for example).  If this is the case, the aim is to simplify the user management: if he writes the `xtypes` namespace, he has access to the public interface of the module. We have chosen this strategy because all xtypes entities are very coupled.

marcoag replied:
I was referring to Message.hpp, as it currently contains an include... maybe just change all includes that point to Message.hpp to xtypes.hpp and remove Message.hpp?

lemunozm replied:
I left the `Message.hpp` file there in order to pack some others inclusions that the messaging might need, or maybe some future utility over the `xtypes` that `soss-core` could offer. The user only needs to carry about the `Message.hpp` inclusion.

marcoag replied:
ok, we can keep it like that then.

marcoag commented on packages/core/src/Config.cpp:
```diff
@@ -429,9 +447,19 @@ bool Config::parse(const YAML::Node& config_node, const std::string& file)
+    {
+        std::cerr << "The 'mock' middleware needs 'types-from' property" << std::endl;
+        return false;
+    }
```
Could we find a better way than hardcoding the "mock" middleware option here?

lemunozm replied:
I will investigate, maybe the own `mock` SH can check itself for this property.

marcoag replied:
I think that would look better, yes.

lemunozm replied:
Already modified.

marcoag reviewed (commented):
After updating the git submodules I get this when trying to colcon build:

` fatal error: dds/core/xtypes/xtypes.hpp: No such file or directory`

What would be the process to build SOSS after this PR is merged?

lemunozm committed:
Minnor changes.

lemunozm commented:
> After updating the git submodules I get this when trying to colcon build:
> 
> ` fatal error: dds/core/xtypes/xtypes.hpp: No such file or directory`
> 
> What would be the process to build SOSS after this PR is merged?

Hi @marcoag ,
The following is enough for me (and avoid to compile packages no updated with xtypes): 
```
colcon build --packages-up-to  soss-local-example
```

When you get the error: ` fatal error: dds/core/xtypes/xtypes.hpp: No such file or directory`, what package is complining?

lemunozm committed:
Updated to xtypes upgrades.

lemunozm committed:
soss-json updates with xtypes.

lemunozm committed:
Removed unused namespace. Updated xtypes thirdparty.

marcoag commented:
Maybe because you have fastrtps installed on your system? (I understand this shouldn't be a requirement for compiling soss)

```
marco@marco:~/workspaces$ colcon build --packages-up-to soss-local-example
Starting >>> fastcdr 
Starting >>> foonathan_memory_vendor
Finished <<< fastcdr [0.23s]                                                                                 
Finished <<< foonathan_memory_vendor [0.24s]
Starting >>> fastrtps
Finished <<< fastrtps [0.26s]                       
Starting >>> soss-core
--- stderr: soss-core                             
In file included from /home/marco/workspaces/soss/src/soss/packages/core/include/soss/SystemHandle.hpp:23:0,
                 from /home/marco/workspaces/soss/src/soss/packages/core/src/register_system.hpp:21,
                 from /home/marco/workspaces/soss/src/soss/packages/core/src/register_system.cpp:18:
/home/marco/workspaces/soss/src/soss/packages/core/include/soss/Message.hpp:21:10: fatal error: dds/core/xtypes/xtypes.hpp: No such file or directory
 #include <dds/core/xtypes/xtypes.hpp>
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
make[2]: *** [CMakeFiles/soss-core.dir/src/register_system.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
In file included from /home/marco/workspaces/soss/src/soss/packages/core/include/soss/SystemHandle.hpp:23:0,
                 from /home/marco/workspaces/soss/src/soss/packages/core/src/register_system.hpp:21,
                 from /home/marco/workspaces/soss/src/soss/packages/core/src/Config.hpp:21,
                 from /home/marco/workspaces/soss/src/soss/packages/core/src/Instance.cpp:18:
/home/marco/workspaces/soss/src/soss/packages/core/include/soss/Message.hpp:21:10: fatal error: dds/core/xtypes/xtypes.hpp: No such file or directory
 #include <dds/core/xtypes/xtypes.hpp>
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
make[2]: *** [CMakeFiles/soss-core.dir/src/Instance.cpp.o] Error 1
In file included from /home/marco/workspaces/soss/src/soss/packages/core/include/soss/SystemHandle.hpp:23:0,
                 from /home/marco/workspaces/soss/src/soss/packages/core/src/register_system.hpp:21,
                 from /home/marco/workspaces/soss/src/soss/packages/core/src/Config.hpp:21,
                 from /home/marco/workspaces/soss/src/soss/packages/core/src/Config.cpp:19:
/home/marco/workspaces/soss/src/soss/packages/core/include/soss/Message.hpp:21:10: fatal error: dds/core/xtypes/xtypes.hpp: No such file or directory
 #include <dds/core/xtypes/xtypes.hpp>
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
make[2]: *** [CMakeFiles/soss-core.dir/src/Config.cpp.o] Error 1
make[1]: *** [CMakeFiles/soss-core.dir/all] Error 2
make: *** [all] Error 2
---
Failed   <<< soss-core	[ Exited with code 2 ]

Summary: 3 packages finished [0.77s]
  1 package failed: soss-core
  1 package had stderr output: soss-core
  2 packages not processed
```

lemunozm commented:
> colcon build --packages-up-to soss-local-example

We updated `xtypes` with some `cmake` modifications the last week. Are you testing with the last commit?

lemunozm commented:
Could you update your `soss_v3_devel` branch to `master`? We are developing our `xtypes-dds` branch over some features that `master` branch offers.

mxgrey commented:
It seems that even with the target branch completely up to date with master, git still considers some of the files to be in conflict. I'm not sure what can be done besides manually resolving the merge conflicts.

mxgrey commented:
Maybe if you merge the current `master` branch into your remote `eProsima/soss_v2:feature/xtypes-dds` branch, that could automatically resolve the conflicts?

lemunozm committed:
Updated xtypes thirdparty.

lemunozm commented:
> Maybe if you merge the current `master` branch into your remote `eProsima/soss_v2:feature/xtypes-dds` branch, that could automatically resolve the conflicts?

Hi Grey,
yes, we were waiting for your rebase. 
Out branch is already rebased too.
Thanks.

marcoag commented:
It seems that `colcon build --packages-up-to  soss-local-example` works now, however when building the rest `soss-ros2` still shows the error pointing out that the `xtypes.hpp` file is not found, maybe we should update the packages that need this so we don't break the overall build.

lemunozm commented:
> It seems that `colcon build --packages-up-to soss-local-example` works now, however when building the rest `soss-ros2` still shows the error pointing out that the `xtypes.hpp` file is not found, maybe we should update the packages that need this so we don't break the overall build.

At this moment, `soss-ros2` is not ported to *xtypes* (the same happens with `soss-websocket` package). We are working now into adding *xtypes* to `soss-ros2`.

lemunozm committed:
Remove fastrtps dependencies: xtypes standalone.

lemunozm committed:
'types-from' mock requirement checked inside of mock package.

lemunozm committed:
Notice the user when a QoS is enabled in a conversion.

lemunozm committed:
xtypes updated.

Commit:
Refs #6853 Core utils. (#8)

lemunozm committed:
Feature/idl yaml (#10)

* First approach to load configure types with idl in yaml.

* Fixed idl-yaml parser.

Commit:
Refs #6853 String template issues fiexd. (#11)

* Refs #6853 String template issues fiexd.

* Requested changes

Commit:
Feature/websocket (#12)

* Refs #6853 Websocket and its compiling using xtypes. Tests fails due to lack of Type registration.

* Refs #6853 Usage update

* Refs #6853 Parser API update

* Refs #6853 Websocket tests working.

* Requested changes

* Tabs inferno!

lemunozm committed:
soss-ros2 xtypes support (#9)

* wip-commit

* rollback for windows support purpose

* indentation mods in MetaPublisher.hpp

* win-supp back on

* ros2-test compiling with xtypes

* wip

* Compiling with xtypes. Empty implementations.

* Architecture for create a ros2 type.

* Convertion tries. wip.

* Compiling with convert.

* Minnor refactor for Convert container.

* Convert working with containers.

* Adapt names. Fixed linker problem.

* Added IDL parser.

* Fixed previous modified style.

* Added domain test working.

* Ros2 tests successful for pub/sub

* Added test_msgs working with xtypes.

* Ported empy for services.

* Fixed some issues with services.

* Passing all ros2 tests with services.

* Revert StringTemplate

* Update packages/ros2/src/MetaPublisher.cpp

Co-Authored-By: Luis Gasco <luisgasco@eprosima.com>

* Fixed required for PR #9

lemunozm committed:
Modified local-example. Now called echo. (#13)

lemunozm committed:
types-from is no longer necessary for the mock.

lemunozm committed:
Updated thirdparty

lemunozm committed:
Clean ros2 prototype. Fixed config required types. (#14)

lemunozm committed:
Ros2 generated idl raw string protected.

lemunozm committed:
Done some todo related with the services. (#15)

lemunozm committed:
Feature/update doc (#17)

* Updated concept.md file.

* Minnor doc changes.

lemunozm committed:
Updated thirdparty.

Commit:
Refs #6981 Added "paths" to types. Modifies the whole "types" section in the yaml, so all examples and tests should be adapted. (#18)

lemunozm committed:
Added ros2 documentation design. (#19)

* Added ros2 documentation design.

* Updated thirdparty.

jamoralp committed:
Add template specialization for Convert<boost::array> utility (#20)

* Add template specialization for Convert<boost::array> utility

* Unify std::array and boost::array template specialization

* Small change to ::array conversion template

jamoralp committed:
Update thirdparty xtypes (#21)

Commit:
Update xtypes package. (#22)

Commit:
Added config node for topics and services. [7780] (#23)

* Added config node for topics and services.

* Added 'to' and 'client'.

* Requested changes

richiware committed:
Refs #7849. Auto-import xtypes when using soss-core. (#24)

richiware committed:
Support xtypes with scoped names. [7969] (#26)

* Refs #7849. Support xtypes with scoped names.

* Update packages/core/src/Config.cpp

Co-Authored-By: Luis Gasco <luisgasco@eprosima.com>

Co-authored-by: Luis Gasco <luisgasco@eprosima.com>

Commit:
Throws an exception if the parser fails. [8057] (#27)

* Throws an exception if the parser fails.

* Fix copy-paste error...

richiware committed:
Support for DDS-RPC methods. [8304] (#28)

* Added resolve member types.

* Fix unimplemented method.

* Added support to serve a same Service with multiple types.

* Fix pointer by value being invalid in lambda.

* Refs #8188. Register soss-ros2 ClientProxy with request type name

* Reverted request-response change and added create_client_proxy with request+reply types.

* Changes default behavior without request/reply in server proxies.

* Fix ros2 server

* Refs #8304. Fixed ROS2 tests.

* Refs #8304. Fixed compilation error on Jenkins.

Co-authored-by: LuisGasco <luisgasco@eprosima.com>
Co-authored-by: Jose Antonio Moral <joseantoniomoralparras@gmail.com>

Commit:
Refs #8052 Adds both forms of scoped name to the types map. (#29)

jamoralp committed:
Refs #8311. Truly generate empty structs for ROS2 SOSS IDL types (#30)

* Refs #8311. Truly generate empty structs for ROS2 SOSS IDL types

* Update xtypes submodule to latest version

Commit:
Avoid infinite waiting if a request don't reach the remote service. (#31)

Commit:
Update submodule (#32)

marcoag commented on .gitmodules:
```diff
@@ -0,0 +1,3 @@
+[submodule "thirdparty/xtypes"]
+	path = thirdparty/xtypes
+	url = https://github.com/eProsima/xtypes.git
```
Might be more robust to have this dependency as a CMake External Project and download it automatically.

marcoag commented on packages/core/include/soss/Message.hpp:
Is there a need for this file anymore? 

In the same line, Messages.cpp might not be necessary either...

marcoag commented on packages/core/CMakeLists.txt:
```diff
@@ -55,6 +55,7 @@ target_include_directories(soss-core
   PUBLIC
     $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>
     $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}>
+    #$<INSTALL_INTERFACE:${xtypes_INCLUDE_DIR}>  #propagate the xtypes headers
```
```suggestion
```
Can I assume we don't need this?

marcoag commented on packages/core/include/soss/SystemHandle.hpp:
```diff
@@ -72,17 +77,20 @@ class SystemHandle
   ///   The configuration specific to this system handle, as described in the
   ///   user-provided yaml input file.
   ///
+  /// \param[in] type_register
```
```suggestion
  /// \param[in] type_registry
```
Although maybe type_registry a more specific name for this parameter can add more value

marcoag commented on packages/echo/example/main.cpp:
```diff
@@ -0,0 +1,56 @@
+
+int main()
+{
+    soss::InstanceHandle soss_handle = soss::run_instance("install/soss-echo/bin/config.yaml");
```
maybe better not to hardcode paths, shall we put the config file as a parameter?

marcoag commented on packages/core/src/MiddlewareInterfaceExtension.cpp:
```diff
@@ -27,6 +27,7 @@
 #else
 #include <dlfcn.h>
 #endif
+#include <assert.h>
```
```suggestion
```

I'm not sure if this include is needed, on Linux is not needed for a successful colcon compilation. In any case either this or the one on line 26 but not both I guess.

marcoag commented on packages/mock/CMakeLists.txt:
```diff
@@ -2,6 +2,7 @@ cmake_minimum_required(VERSION 3.5.0)
 
 project(soss-mock)
 
+find_package(xtypes REQUIRED) #why?
```
Doesn't seem to be needed, however maybe this line should remain if this mock plugin is supposed to serve as an example... shouldn't it?

marcoag commented:
Branch has been updated to master and there seems to be some conflicts, @lemunozm could you guys resolve this on your side?

marcoag commented on packages/core/src/Config.cpp:
```diff
@@ -107,14 +111,97 @@ std::unique_ptr<ServiceRoute> parse_service_route(const YAML::Node& node)
+
+  if (!node["types"]["idls"])
+  {
+    std::cerr << "The config file [" << filename << "] has a 'types' entry which doesn't contains an 'idls' entry."
```
```suggestion
    std::cerr << "The config file [" << filename << "] has a 'types' entry "
              << "which doesn't contains an 'idls' entry."
```

marcoag commented on packages/core/src/Config.cpp:
```diff
@@ -107,14 +111,97 @@ std::unique_ptr<ServiceRoute> parse_service_route(const YAML::Node& node)
+      if (types.empty())
+      {
+          std::cout << "Parsing the idl '" << idl_index << "' placed in the yaml config. "
+                    << "The parsing was successful but no types were found."
```
```suggestion
          std::cout << "Parsing the idl '" << idl_index << "' placed in the "
                    << "yaml config. The parsing was successful but no "
                    <<"types were found."
```

marcoag commented on packages/core/src/Config.cpp:
```diff
@@ -107,14 +111,97 @@ std::unique_ptr<ServiceRoute> parse_service_route(const YAML::Node& node)
+    else
+    {
+      std::cerr << "Error parsing the idl '" << idl_index << "' placed in the yaml config. "
+                << "Please, review and fix your idl specification syntax."
```
```suggestion
          std::cerr << "Error parsing the idl '" << idl_index << "' placed in "
                << "the yaml config. Please, review and fix your idl "
                << "specification syntax."
```

marcoag commented on packages/core/src/Config.cpp:
```diff
@@ -172,38 +292,59 @@ bool add_topic_or_service_config(
+      if (!request_type && !reply_type)
+      {
+        std::cerr << channel_type << " configuration [" << name
+                  << "] is missing its [type], or [request_type] and [reply_type] fields!" << std::endl;
```
```suggestion
        std::cerr << channel_type << " configuration [" << name
                  << "] is missing its [type], or [request_type] and "
                  << "[reply_type] fields!" << std::endl;
```

marcoag commented on packages/core/src/Config.cpp:
```diff
@@ -411,96 +613,162 @@ bool Config::parse(const YAML::Node& config_node, const std::string& file)
+      if (m_middlewares.find(it_remap.first) == m_middlewares.end())
+      {
+        std::cerr << "Unrecognized system [" << it_remap.first << "] requested for mapping topic "
+                  << "[" << entry.first << "]" << std::endl;
```
```suggestion
            std::cerr << "Unrecognized system [" << it_remap.first
                  << "] requested for mapping topic ["
                  << entry.first << "]" << std::endl;
```

marcoag commented on packages/core/src/Config.cpp:
```diff
@@ -411,96 +613,162 @@ bool Config::parse(const YAML::Node& config_node, const std::string& file)
+      if (m_middlewares.find(it_remap.first) == m_middlewares.end())
+      {
+        std::cerr << "Unrecognized system [" << it_remap.first << "] requested for mapping service "
+                  << "[" << entry.first << "]" << std::endl;
```
```suggestion
        std::cerr << "Unrecognized system [" << it_remap.first 
                  << "] requested for mapping service ["
                  << entry.first << "]" << std::endl;
```

marcoag commented on packages/core/src/Config.cpp:
```diff
@@ -554,47 +825,133 @@ bool Config::load_middlewares(SystemHandleInfoMap& info_map) const
+        const auto it = info_map.find(type);
+        if (it == info_map.end())
+        {
+          std::cerr << "'types-from' references to a non-existant middleware" << std::endl;
```
```suggestion
         std::cerr << "'types-from' references to a non-existant middleware"
                   << std::endl;
```

marcoag commented on packages/core/src/Config.cpp:
```diff
@@ -554,47 +825,133 @@ bool Config::load_middlewares(SystemHandleInfoMap& info_map) const
+        if (!types.count(required_type))
+        {
+          std::cerr << "The middleware '" << mw_name << "' must satisfy the required "
+                    << "type '" << required_type << "'"
```
```suggestion
          std::cerr << "The middleware '" << mw_name << "' must satisfy the "
                    << "required type '" << required_type << "'"
```

marcoag commented on packages/core/src/Config.cpp:
```diff
@@ -554,47 +825,133 @@ bool Config::load_middlewares(SystemHandleInfoMap& info_map) const
+        if (!types.count(required_type))
+        {
+          std::cerr << "The middleware '" << mw_name << "' must satisfy the required "
+                    << "type '" << required_type << "'"
```
```suggestion
          std::cerr << "The middleware '" << mw_name << "' must satisfy the "
                    << "required type '" << required_type << "'"
```

marcoag commented on packages/core/src/Config.cpp:
```diff
@@ -673,32 +1073,53 @@ bool Config::configure_services(const SystemHandleInfoMap& info_map) const
-          remap_if_needed(server, config.remap, service_name),
-          config.service_type,
-          config_or_empty_node(server, config.middleware_configs));
+    TopicInfo server_info = remap_if_needed(server, config.remap, {service_name, config.request_type, config.reply_type});
```
```suggestion
    TopicInfo server_info = remap_if_needed(server, 
                                            config.remap, 
                                            {service_name, 
                                             config.request_type, 
                                             config.reply_type});
```

marcoag commented on packages/core/src/Config.cpp:
```diff
@@ -673,32 +1073,53 @@ bool Config::configure_services(const SystemHandleInfoMap& info_map) const
-          config_or_empty_node(server, config.middleware_configs));
+    TopicInfo server_info = remap_if_needed(server, config.remap, {service_name, config.request_type, config.reply_type});
+    const xtypes::DynamicType* server_type = resolve_type(it->second.types, server_info.type);
+    //const xtypes::DynamicType& server_type = *it->second.types.find(server_info.type)->second;
```
```suggestion
```
maybe?

marcoag commented on packages/core/src/Config.cpp:
```diff
@@ -673,32 +1073,53 @@ bool Config::configure_services(const SystemHandleInfoMap& info_map) const
-                << server << "] for service type [" << config.service_type
-                << "]" << std::endl;
-      return false;
+      const xtypes::DynamicType* server_reply_type = resolve_type(it->second.types, server_info.reply_type);
```
```suggestion
      const xtypes::DynamicType* server_reply_type = 
          resolve_type(it->second.types, 
                       server_info.reply_type);
```

marcoag commented on packages/core/src/Config.cpp:
```diff
@@ -707,15 +1128,229 @@ bool Config::configure_services(const SystemHandleInfoMap& info_map) const
+      TopicInfo client_info = remap_if_needed(client, config.remap,
+                                              {service_name, config.request_type, config.reply_type});
+      const xtypes::DynamicType* client_type = resolve_type(it->second.types, client_info.type);
+      //const xtypes::DynamicType& client_type = *it->second.types.find(client_info.type)->second;
```
```suggestion
```
maybe?

marcoag commented on packages/core/src/Config.cpp:
```diff
@@ -707,15 +1128,229 @@ bool Config::configure_services(const SystemHandleInfoMap& info_map) const
+      {
+        valid &= it->second.service_client->create_client_proxy(
+          client_info.name,
+          //*client_type,
```
```suggestion
```
maybe?

marcoag commented on packages/core/src/Config.cpp:
```diff
@@ -707,15 +1128,229 @@ bool Config::configure_services(const SystemHandleInfoMap& info_map) const
+
+        valid &= it->second.service_client->create_client_proxy(
+          client_info.name,
+          //*client_type,
```
```suggestion
```
?

marcoag commented on packages/core/src/Config.cpp:
```diff
@@ -707,15 +1128,229 @@ bool Config::configure_services(const SystemHandleInfoMap& info_map) const
+            {
+              provider->call_service(request, client, call_handle);
+            }
+            else             //previously ensured that TypeConsistency is not NONE
```
```suggestion
            else //previously ensured that TypeConsistency is not NONE
```

marcoag commented on packages/core/src/Config.cpp:
```diff
@@ -707,15 +1128,229 @@ bool Config::configure_services(const SystemHandleInfoMap& info_map) const
+      }
+      */
+
+      //xtypes::TypeConsistency consistency = from_type->second->is_compatible(*to_type->second);
```
```suggestion

    for (const std::string& to : config.route.to)
    {
      const auto it_to = info_map.find(to);
      TopicInfo topic_info_to = remap_if_needed(to, 
                                                config.remap, 
                                                {topic_name, 
                                                 config.message_type, ""});
      const xtypes::DynamicType* to_type = resolve_type(it_from->second.types,
                                                        topic_info_from.type);
```
Shall we remove the commented code?

marcoag commented on packages/core/src/Config.cpp:
```diff
@@ -707,15 +1128,229 @@ bool Config::configure_services(const SystemHandleInfoMap& info_map) const
+    }
+    */
+
+    //if (client_type->second->is_compatible(*server_type->second) == xtypes::TypeConsistency::NONE)
```
```suggestion

    const auto it_server = info_map.find(config.route.server);
    TopicInfo topic_info_server = remap_if_needed(config.route.server, 
                                                  config.remap,
                                                  {service_name, 
                                                   config.request_type,
                                                   config.reply_type});
    const xtypes::DynamicType* server_type = 
        resolve_type(it_server->second.types, topic_info_server.type);
```
Shall we remove the commented code?

marcoag commented on packages/ros2-test/integration/ros2__test_msgs.cpp:
```diff
@@ -163,351 +150,147 @@ TEST_CASE("Transmit and receive all test messages", "[ros2]")
-    msg.primitive_values[i] = generate_random_primitives(i);
+    for (std::size_t i = 0; i < N; ++i)
+    {
+        CHECK(received_array.primitive_values[i] == ros2_msg.primitive_values[i]);
```
```suggestion
        CHECK(received_array.primitive_values[i] 
                == ros2_msg.primitive_values[i]);
```

jamoralp committed:
Support for ROS2 Foxy (#33)

* Cross-compatibility with eloquent and foxy (#44)

(cherry picked from commit 2ee5bc55f4c21e59a193f63ace7521366f137597)

* Fix SOSS-ROS2 compilation with colcon

* Remove unneeded boost dependencies

Co-authored-by: Grey <grey@openrobotics.org>

russkel commented:
Hi, is this still under development? I am quite interested in the xtypes features for a project.

mxgrey commented:
Due to a shortage of resources that Open Robotics can commit to SOSS development, I believe eProsima intends to take over as the maintainers of the SOSS/Integration-Service project. Perhaps @lemunozm could advise on where to find the latest developments.

lemunozm commented:
Hi @russkel, @mxgrey, I am no longer working on this project, sorry, I can't help

russkel commented:
Ah that's unfortunate this was never merged in. Thanks for the update.

jamoralp committed:
Update xtypes submodule (#39)

lauramg15 committed:
Fix websocket tests (#40)

lauramg15 committed:
Skip blank Services' names (#38)

jamoralp committed:
Fix soss-ros2 tests, add information traces (#41)

* Skip blank Services' names

* Fix soss-ros2 tests

Co-authored-by: laura-eprosima <laura@eprosima.com>

jacobperron committed:
Fix compile error (#37)

Include <cassert> header to resolve the following error:

    MiddlewareInterfaceExtension.cpp:98:5: error: ‘assert’ was not declared in this scope
       98 |     assert(_directory.is_absolute());
          |     ^~~~~~

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

mathias-luedtke committed:
Fix types_from passing (#34)

* sort middlewares by types_from hierarchy

* provide "types_from" to dependent system handles during configuration

mathias-luedtke committed:
Add handling for special float/double values (#35)

* Add handling for special float/double values

* JSON cannot express inf, -inf, nan, -nan
* Common (non-compliant) work-around: store as string

* Code cleanup and styling

Co-authored-by: Jose Antonio Moral <joseantoniomoralparras@gmail.com>

mathias-luedtke committed:
Copy received message in soss-echo before popping it (#36)

lauramg15 committed:
Uncrustify (#44)

* Apply uncrustify to Core

* Apply uncrustify to Echo System Handle

* Apply uncrustify to Json System Handle

* Apply uncrustify to Mock System Handle

* Apply uncrustify to ROS2 System Handle

* Apply uncrustify to Websocket System Handle

* Apply uncrustify to examples

* Apply uncrutify to Core tests

* Solve uncrustify issues

lauramg15 committed:
Fixes crash due to exception while closing the websocket connections (#42)

lauramg15 committed:
Adds support for YAML request_type and reply_type tags (#43)

lauramg15 committed:
Fixes the use of types-from option for the WebSocket System Handle (#46)

lauramg15 committed:
New test for checking websocket client-server interaction (#45)

* New test for checking websocket client-server interaction

* Apply suggestions

jamoralp commented:
> Ah that's unfortunate this was never merged in. Thanks for the update.

Good morning @russkel, as part of the team in eProsima responsible of this project I am going to be mainly involved in transferring the repository to our organization and updating the main branch with the xTypes implementation. Keep an eye on it, it should be available in the following months. I will keep you updated!

russkel commented:
Hi @jamoralp, I saw the new commits to the repository and was relieved! Thanks for the message and your work on it.

jamoralp commented:
New SOSS_v3 release will be named *Integration Service* and will feature the whole set of functionalities added in this pull request regarding `xtypes`. The `feature/xtypes-dds` branch will be merged onto a brand new `main` branch by april. Therefore, we can close this
